### PR TITLE
Support getting dependencies info for a test

### LINKED ISSUE #893: TestNG should provide an Api which allow to find all dependent of a specific test
https://issues.apache.org/jira/browse/SUREFIRE-1073

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #893

Fixes #893  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`
- [X] Auto applied styling via `./gradlew autostyleApply`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

**Note:** For more information on contribution guidelines  please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

### CODE REVIEW

File: testng-core/src/main/java/org/testng/internal/BaseTestMethod.java [hunk 1]
```diff
@@ -176,6 +178,34 @@ public String[] getMethodsDependedUpon() {
...
+    if (RuntimeBehavior.isMemoryFriendlyMode()) {
+      downstreamDependencies =
+          methods.stream().map(LiteWeightTestNGMethod::new).collect(Collectors.toSet());
+    } else {
+      downstreamDependencies.addAll(methods);
+    }
+  }
+
+  public void setUpstreamDependencies(Set<ITestNGMethod> methods) {
+    if (RuntimeBehavior.isMemoryFriendlyMode()) {
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/internal/BaseTestMethod.java:201]: 👍

File: testng-core/src/test/java/test/dependent/DependentTest.java [hunk 2]
```diff
@@ -218,6 +222,72 @@ public void testMethodDependencyAmidstInheritance() {
...
+  public void testDownstreamDependencyRetrieval(
+      Class<?> clazz, String independentMethod, String[] dependentMethods) {
+    TestNG testng = create(clazz);
+    DependencyTrackingListener listener = new DependencyTrackingListener();
+    testng.addListener(listener);
+    testng.run();
+    String cls = clazz.getCanonicalName();
+    String key = cls + "." + independentMethod;
+    Set<String> downstream = listener.getDownstreamDependencies().get(key);
+    if (dependentMethods == null) {
```
Comment (juherr) [inline testng-core/src/test/java/test/dependent/DependentTest.java:235]: Maybe replace null by empty array. The test will easier to read

**Review by juherr — CHANGES_REQUESTED** [review]
Except the default behavior of interfaces, everything else is awesome.

**Comment by krmahadevan** [comment]
@juherr - Fixed all the review comments, amended my commit and pushed the changes. Please help take a look.

File: testng-core-api/src/main/java/org/testng/ITestNGMethod.java [hunk 3]
```diff
@@ -71,6 +73,24 @@ public interface ITestNGMethod extends Cloneable {
    */
   String[] getMethodsDependedUpon();
 
+  /**
+   * @return - The set of methods that are dependent on the current method. This information can
+   *     help in deciding what other TestNG methods will be skipped if the current method fails. If
+   *     the current method is a configuration method, then an empty set is returned.
+   */
+  default Set<ITestNGMethod> downstreamDependencies() {
+    return Collections.emptySet();
```
Comment (juherr) [inline testng-core-api/src/main/java/org/testng/ITestNGMethod.java:82]: Exception please

File: testng-core-api/src/main/java/org/testng/ITestNGMethod.java [hunk 4]
```diff
@@ -71,6 +73,24 @@ public interface ITestNGMethod extends Cloneable {
...
+  }
+
+  /**
+   * @return - The set of methods upon which the current method has a dependency. This information
+   *     can help in deciding what all TestNG methods need to pass before the current method can be
+   *     executed. If the current method is a configuration method, then an empty set is returned.
+   */
+  default Set<ITestNGMethod> upstreamDependencies() {
+    return Collections.emptySet();
```
Comment (juherr) [inline testng-core-api/src/main/java/org/testng/ITestNGMethod.java:91]: Exception please

**Review by juherr — CHANGES_REQUESTED** [review]

**Comment by krmahadevan** [comment]
@juherr - Please check now. default implementation has been flipped to throwing an exception

File: testng-core/src/main/java/org/testng/internal/BaseTestMethod.java [hunk 5]
```diff
@@ -176,6 +178,34 @@ public String[] getMethodsDependedUpon() {
...
+  public Set<ITestNGMethod> upstreamDependencies() {
+    return Collections.unmodifiableSet(upstreamDependencies);
+  }
+
+  public void setDownstreamDependencies(Set<ITestNGMethod> methods) {
+    if (RuntimeBehavior.isMemoryFriendlyMode()) {
+      downstreamDependencies =
+          methods.stream().map(LiteWeightTestNGMethod::new).collect(Collectors.toSet());
+    } else {
+      downstreamDependencies.addAll(methods);
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/internal/BaseTestMethod.java:196]: If the method is called more than once, the behavior won't be the same in all cases: sometimes the first setup is lost and sometimes the setup is added.

Just clear before add or create a new set.

Same remark for upstream dep.

**Review by juherr — CHANGES_REQUESTED** [review]

File: testng-core/src/main/java/org/testng/internal/BaseTestMethod.java (hunk 5 above)
Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/internal/BaseTestMethod.java:199]: This is not going to happen. The upstream/downstream dependencies are computed and injected into a test method even before the execution happens. So this setting is called ONLY once per `<test>` tag. But I have still added the edit check to clear the set if its not empty before setting upstream/downstream dependencies.
Please check now

File: testng-core/src/main/java/org/testng/internal/BaseTestMethod.java [hunk 6]
```diff
@@ -176,6 +178,40 @@ public String[] getMethodsDependedUpon() {
...
+    }
+  }
+
+  public void setUpstreamDependencies(Set<ITestNGMethod> methods) {
+    if (!upstreamDependencies.isEmpty()) {
+      upstreamDependencies.clear();
+    }
+    if (RuntimeBehavior.isMemoryFriendlyMode()) {
+      upstreamDependencies =
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/internal/BaseTestMethod.java:208]: You should be able to addAll here too and make the sets final

**Review by juherr — APPROVED** [review]

**Pull request merged**